src/structs.py

```diff
@@ -239,6 +240,25 @@ def __str__(self) -> str:
     def __repr__(self) -> str:
         return str(self)
 
+    def pretty_str(self) -> Tuple[str, str]:
+        """Display the predicate in a nice human-readable format.
+
+        Returns a tuple of (variables string, body string).
+        """
+        if hasattr(self._classifier, "pretty_str"):
+            # This is an invented predicate, from the predicate grammar.
+            pretty_str_f = getattr(self._classifier, "pretty_str")
```

Comment:
hack around mypy which was complaining if you just do self._classifier.pretty_str() bc self._classifier might be a function. tried other ways to get around this, but this was easiest